pkg/controller/logstorage/elastic/elastic_controller.go

```diff
@@ -0,0 +1,710 @@
+	}
+
+	if err = utils.AddSecretsWatch(c, render.ElasticsearchCuratorUserSecret, common.OperatorNamespace()); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch Secret resource: %w", err)
+	}
+
+	if err = utils.AddConfigMapWatch(c, relasticsearch.ClusterConfigConfigMapName, common.OperatorNamespace(), &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch ConfigMap resource: %w", err)
+	}
+
+	if err := utils.AddServiceWatch(c, render.ElasticsearchServiceName, render.ElasticsearchNamespace); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch the Service resource: %w", err)
+	}
+
+	if err := utils.AddServiceWatch(c, esgateway.ServiceName, render.ElasticsearchNamespace); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch the Service resource: %w", err)
+	}
+
+	if err := utils.AddServiceWatch(c, render.LinseedServiceName, render.ElasticsearchNamespace); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch the Service resource: %w", err)
+	}
+
+	// Watch for changes to primary resource ManagementCluster
+	err = c.Watch(&source.Kind{Type: &operatorv1.ManagementCluster{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch primary resource: %w", err)
+	}
+
+	// Watch for changes to primary resource ManagementClusterConnection
+	err = c.Watch(&source.Kind{Type: &operatorv1.ManagementClusterConnection{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch primary resource: %w", err)
+	}
+
+	return nil
+}
+
+func (r *ElasticSubController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling LogStorage - Elasticsearch")
+
+	// Get LogStorage resource.
+	ls := &operatorv1.LogStorage{}
+	key := utils.DefaultTSEEInstanceKey
+	err := r.client.Get(ctx, key, ls)
+	if err != nil {
+		// Not finding the LogStorage CR is not an error, as a Managed cluster will not have this CR available but
+		// there are still "LogStorage" related items that need to be set up
+		// TODO: This controller should be responsible for installing an ES cluster and nothing more.
+		// Any extra bits should be moved to companion controllers to reduce code complexity.
+		if !errors.IsNotFound(err) {
+			r.status.SetDegraded(operatorv1.ResourceReadError, "An error occurred while querying LogStorage", err, reqLogger)
+			return reconcile.Result{}, err
+		}
+		ls = nil
+		r.status.OnCRNotFound()
+	} else {
+		// We found the LogStorage instance.
+		r.status.OnCRFound()
+	}
+
+	// Wait for the initializing controller to indicate that the LogStorage object is actionable.
+	if ls == nil || ls.Status.State != operatorv1.TigeraStatusReady {
```

Comment:
For managed clusters (that have no logstorage), would this message be misleading? And probably we would not want to requeue

Reply:
@rene-dekker yeah, this is definitely a bit funky.

I am wondering if we should decouple the management / managed implementation so that this controller can confidently short-circuit if there is no LogStorage. That seems like the cleanest implementation to me, and is less likely to result in accidentally writing bugs into this controller.